tests/tests/wgpu-validation/api/error_scopes.rs

```diff
@@ -1,42 +1,116 @@
     // Register an uncaptured error handler to catch errors from other threads.
-    device.on_uncaptured_error(Arc::new(move |_error| {
-        other_thread_error_clone.store(true, Ordering::Relaxed);
-    }));
+    let other_thread_error = register_uncaptured_error_handler(&device);
 
     // Do something invalid on another thread.
     std::thread::scope(|s| {
         s.spawn(|| {
-            let _buffer = device.create_buffer(&wgpu::BufferDescriptor {
-                label: None,
-                size: 1 << 63, // Too large!
-                usage: wgpu::BufferUsages::COPY_SRC,
-                mapped_at_creation: false,
-            });
+            raise_validation_error(&device);
         });
     });
 
     // Pop the error scope on the main thread.
-    let error = pollster::block_on(device.pop_error_scope());
+    let error = pollster::block_on(scope.pop());
 
     // The main thread's error scope should not have captured the other thread's error.
     assert!(error.is_none());
     // The other thread's error should have been reported to the uncaptured error handler.
     assert!(other_thread_error.load(Ordering::Relaxed));
 }
+
+// Test that error scopes error when popped in the wrong order.
+#[test]
+#[should_panic(expected = "error scopes must be popped in reverse order")]
+fn pop_out_of_order() {
+    let (device, _queue) = wgpu::Device::noop(&wgpu::DeviceDescriptor::default());
+
+    let scope1 = device.push_error_scope(wgpu::ErrorFilter::Validation);
+    let _scope2 = device.push_error_scope(wgpu::ErrorFilter::Validation);
+
+    let _ = pollster::block_on(scope1.pop());
+}
+
+// Test that error scopes are automatically popped when dropped.
+#[test]
+fn drop_automatically_pops() {
+    let (device, _queue) = wgpu::Device::noop(&wgpu::DeviceDescriptor::default());
+
+    let uncaptured_error = register_uncaptured_error_handler(&device);
+
+    let scope = device.push_error_scope(wgpu::ErrorFilter::Validation);
+    raise_validation_error(&device);
+    drop(scope); // Automatically pops the error scope.
+
+    assert_eq!(uncaptured_error.load(Ordering::Relaxed), false);
+
+    // Raising another error will go to the uncaptured error handler, not the dropped scope.
+    raise_validation_error(&device);
+
+    assert_eq!(uncaptured_error.load(Ordering::Relaxed), true);
+}
+
+// Test that error scopes are automatically popped when dropped during unwinding,
+// even when they are dropped out of order.
+#[test]
+fn drop_during_unwind() {
+    let (device, _queue) = wgpu::Device::noop(&wgpu::DeviceDescriptor::default());
+
+    let scope1 = device.push_error_scope(wgpu::ErrorFilter::Validation);
+    let scope2 = device.push_error_scope(wgpu::ErrorFilter::Validation);
+
+    let _ = std::panic::catch_unwind(AssertUnwindSafe(|| {
+        raise_validation_error(&device);
+        // Move scope1 so that it is dropped before scope2.
+        let _scope2 = scope2;
+        let _scope1 = scope1;
+        panic!("trigger unwind");
```

Comment:
This could be a `resume_unwind()` to avoid running the panic hook and cluttering test output. (`resume_unwind()` doesn't care if you're actually resuming a prior unwind.) Up to you whether that's a good thing.